### PR TITLE
Add a self test command.

### PR DESCRIPTION
**Pull request by nomis52**
This puts the driver circuit into loopback mode, sends a single byte, then
verifies we received it. That way we can check that the 485 driver is working.

Side note: using separate control lines for TX / RX Enable just paid off!!

### CODE REVIEW

File: firmware/src/app.c [hunk 1]
```diff
@@ -156,9 +156,9 @@ void APP_Tasks(void) {
   USBTransport_Tasks();
   Transceiver_Tasks();
   USBConsole_Tasks();
-  RDMResponder_Tasks();
 
   if (Transceiver_GetMode() == T_MODE_RESPONDER) {
+    RDMResponder_Tasks();
```
Comment (peternewman) [inline firmware/src/app.c]: Haven't we seen this already?

File: firmware/src/transceiver.c [hunk 2]
```diff
@@ -95,6 +108,7 @@ typedef enum {
   OP_RDM_BROADCAST = T_OP_RDM_BROADCAST,
   OP_RDM_WITH_RESPONSE = T_OP_RDM_WITH_RESPONSE,
   OP_RX = T_OP_RX,
+  OP_SELF_TEST = T_OP_SELF_TEST,
   OP_RDM_DUB_RESPONSE,  //!< No break
   OP_RDM_RESEPONSE  //!< With a break
 } InternalOperation;
```
Comment (peternewman) [inline firmware/src/transceiver.c]: Can you Doxygen this enum, or remove the blank @brief so we know it's not been documented.

File: firmware/src/transceiver.c [hunk 3]
```diff
@@ -1332,10 +1424,72 @@ void Transceiver_Tasks() {
       g_transceiver.data_index = 0u;
       g_transceiver.state = STATE_R_RX_PREPARE;
       break;
+
+    // Self Test States
+    case STATE_T_INITIALIZE:
+      PLIB_USART_TransmitterDisable(g_hw_settings.usart);
```
Comment (peternewman) [inline firmware/src/transceiver.c]: Does this need to check the mode first too?

File: firmware/src/transceiver.h [hunk 4]
```diff
@@ -23,26 +23,48 @@
...
+ * receives the same token and any operation-specific data. This allows the
+ * client to associate the callback with the original request that generated
+ * the event.
+ *
+ * If a TRANSCEIVER_NO_NOTIFICATION is used for the token, the callback will
+ * not be run.
+ *
+ * The transceiver can operate in various modes:
+ *  - DMX / RDM Controller
+ *  - DMX / RDM receiver
```
Comment (peternewman) [inline firmware/src/transceiver.h]: SPaG consistency Receiver.

File: firmware/src/transceiver.h [hunk 5]
```diff
@@ -254,11 +284,16 @@ void Transceiver_Initialize(const TransceiverHardwareSettings *settings,
 /**
  * @brief Change the operating mode of the transceiver.
  * @param mode the new operating mode.
+ * @param token to token to the TransceiverEventCallback when the mode change
```
Comment (peternewman) [inline firmware/src/transceiver.h]: SPaG I'm not sure exactly what you're trying to say, all I know is it's not right. :)

File: firmware/src/transceiver.h (hunk 4 above)
Comment (peternewman) [inline firmware/src/transceiver.h]: Done

File: firmware/src/transceiver.h (hunk 5 above)
Comment (peternewman) [inline firmware/src/transceiver.h]: Possibly the second token should be pass?

**Comment by peternewman** [comment]
We're happy to take the risk of a false result if two units already on the same DMX bus run self test at the same time, with one broken Tx?

**Comment by peternewman** [comment]
Generally LGTM

File: firmware/src/transceiver.c (hunk 3 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: We can only end up here from the SwitchMode() function.

File: firmware/src/transceiver.c (hunk 2 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: Removed

File: firmware/src/app.c (hunk 1 above)
Comment (nomis52) [inline firmware/src/app.c]: Yup, synced with master and now this diff has gone.

File: firmware/src/transceiver.h (hunk 5 above)
Comment (nomis52) [inline firmware/src/transceiver.h]: Fixed :)

**Comment by nomis52** [comment]
The self-test is depends on the unit being disconnected from the DMX bus. I'll make a note of that somewhere.

File: firmware/src/transceiver.c (hunk 3 above)
Comment (peternewman) [inline firmware/src/transceiver.c]: Is that a no then @nomis52 ?

**Comment by peternewman** [comment]
LGTM

File: firmware/src/transceiver.c (hunk 3 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: Yeah, we don't need it.

**Pull request merged**